Userland/Libraries/LibWeb/Fetch/Request.cpp

```diff
@@ -0,0 +1,650 @@
+    if (init.cache.has_value())
+        request.set_cache_mode(from_bindings_enum(*init.cache));
+
+    // 21. If request’s cache mode is "only-if-cached" and request’s mode is not "same-origin", then throw a TypeError.
+    if (request.cache_mode() == Infrastructure::Request::CacheMode::OnlyIfCached && request.mode() != Infrastructure::Request::Mode::SameOrigin)
+        return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "Mode must be 'same-origin' when cache mode is 'only-if-cached'"sv };
+
+    // 22. If init["redirect"] exists, then set request’s redirect mode to it.
+    if (init.redirect.has_value())
+        request.set_redirect_mode(from_bindings_enum(*init.redirect));
+
+    // 23. If init["integrity"] exists, then set request’s integrity metadata to it.
+    if (init.integrity.has_value())
+        request.set_integrity_metadata(*init.integrity);
+
+    // 24. If init["keepalive"] exists, then set request’s keepalive to it.
+    if (init.keepalive.has_value())
+        request.set_keepalive(*init.keepalive);
+
+    // 25. If init["method"] exists, then:
+    if (init.method.has_value()) {
+        // 1. Let method be init["method"].
+        auto method = *init.method;
+
+        // 2. If method is not a method or method is a forbidden method, then throw a TypeError.
+        if (!Infrastructure::is_method(method.bytes()))
+            return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "Method has invalid value"sv };
+        if (Infrastructure::is_forbidden_method(method.bytes()))
+            return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "Method must not be one of CONNECT, TRACE, or TRACK"sv };
+
+        // 3. Normalize method.
+        method = MUST(Infrastructure::normalize_method(method.bytes()));
+
+        // 4. Set request’s method to method.
+        request.set_method(MUST(ByteBuffer::copy(method.bytes())));
+    }
+
+    // 26. If init["signal"] exists, then set signal to it.
+    if (init.signal.has_value())
+        input_signal = *init.signal;
+
+    // 27. Set this’s request to request.
+    // NOTE: This is done at the beginning as the 'this' value Request object
+    //       cannot exist with a null Infrastructure::Request.
+
+    // 28. Set this’s signal to a new AbortSignal object with this’s relevant Realm.
+    request_object->m_signal = realm.heap().allocate<DOM::AbortSignal>(HTML::relevant_realm(*request_object), html_window);
+
+    // 29. If signal is not null, then make this’s signal follow signal.
+    if (input_signal != nullptr) {
+        // FIXME: Our AbortSignal doesn't support 'following' yet.
+    }
+
+    // 30. Set this’s headers to a new Headers object with this’s relevant Realm, whose header list is request’s header list and guard is "request".
+    request_object->m_headers = realm.heap().allocate<Headers>(realm, html_window);
+    request_object->m_headers->set_header_list(request.header_list());
```

Comment:
This is correct & the reason why HeaderList is now refcounted - I didn't notice this myself until the very end when I tested after implementing everything :laughing:
The important bit here is "is request’s header list", not "is **a copy of** request’s header list", meaning they point to the same instance. This is so that changes to the JS Headers object are reflected in the Request object's internal infra request and vice versa.

Reply:
Got it, I saw lots of copies but it just didn't click.